pandas/core/arrays/categorical.py

```diff
@@ -1740,6 +1740,17 @@ def take_nd(self, indexer, allow_fill=True, fill_value=None):
         # but is passed thru internally
         assert isna(fill_value)
 
+        indexer = np.asarray(indexer)
+
+        # take on empty array
+        if not len(self):
+            # only valid if result is an all-missing array
+            if (indexer == -1).all():
+                codes = [-1] * len(indexer)
+            else:
+                raise IndexError(
+                    "cannot do a non-empty take from an empty array.")
+
         codes = take_1d(self._codes, indexer, allow_fill=True, fill_value=-1)
```

Comment:
For the internal use cases (Categorical at the moment, later maybe more arrays), I could also push this test to `take_1d`. But, that is also used by many other things, so not sure I can just do that.

Reply:
this should not be here rather in take1d (i see you comment)

Reply:
I simplified the check a bit here (as it only needed to check the empty and not-all -1 case, as when indexer is all -1 `take_1d` works as expected on empty arrays). 

@jreback I am not sure if I should move it to `take_1d`. It depends on the guarantees we want to give (ourselves) for `take_1d` (if it should do such bounds checking). It might be that currently all codes that uses `take_1d/nd` already does this bounds checking (I did not find a practical use case (apart from Categorical.take itself) where we run into this bug), and then doing another one in `take_nd` would give a performance penalty.